packages/react-client/src/ReactFlightReplyClient.js

```diff
@@ -1151,64 +1154,86 @@ export function registerBoundServerReference<T: Function>(
     Object.defineProperties((reference: any), {
       $$FORM_ACTION: {value: $$FORM_ACTION},
       $$IS_SIGNATURE_EQUAL: {value: isSignatureEqual},
-      bind: {value: bind},
     });
+    defineBind(reference);
   }
   knownServerReferences.set(reference, {id, bound});
 }
 
+// TODO: Ideally we'd use `isServerReference` from
+// 'react-server/src/ReactFlightServerConfig', but that can only be imported
+// in a react-server environment.
+function isServerReference(reference: Object): boolean {
+  return reference.$$typeof === Symbol.for('react.server.reference');
+}
+
 export function registerServerReference<T: Function>(
   reference: T,
   id: ServerReferenceId,
   encodeFormAction?: EncodeFormActionCallback,
 ): ServerReference<T> {
-  registerBoundServerReference(reference, id, null, encodeFormAction);
+  const bound =
+    isServerReference(reference) && reference.$$bound
+      ? Promise.resolve(reference.$$bound)
```

Comment:
We can't make these assumptions about the structure of a Server Reference here. It's up to the bundler.

In fact, these should not even be expandos. I've been meaning to move all the fields on Server/Client References onto WeakMaps since they're internal state owned by the server and should be opaque to the surrounding environment - including Flight's Client.

Reply:
Also, if it is bound already then it's used incorrectly and should be an error (if it was even possible to check).

Because `registerServerReference` should always be called on the original function, not the closure. It represents the raw function and not the closure.

I.e. to create a bound function you need to pass the original function to both both `registerServerReference`s before binding it.

```
ReactFlightClient.registerServerReference(ReactFlightServer.registerServerReferenace(fn, ...), ...).bind(...)

```